app/src/auth.py

```diff
@@ -49,50 +54,63 @@ def login_post():
 def register():
     """Register user"""
-
+ 
     if request.method == "POST":
 
         name = request.form.get("name")
         username = request.form.get("username")
         date_of_birth = request.form.get("date_of_birth")
-        grade = request.form.get("grade")
+        grade = getattr(GradeEnum, request.form.get("grade"))   # Parses string into GradeEnum field and assigns proper grade
         email = request.form.get("email")
+        confirm_email = request.form.get("confirm_email")
         password = request.form.get("password")
         confirm_password = request.form.get("confirm_password")
 
+        age = calculate_age(date_of_birth)
         user_email = User.query.filter_by(email=email).first()
-        if user_email:
-            flash("This email already exists.")
-            return redirect(url_for("auth.register"))
         user_name = User.query.filter_by(username=username).first()
+
+        if len(name) == 0:
+            flash("You must provide your name.", "error_name")
+            return redirect(url_for("auth.register"))
         if user_name:
-            flash("This username already exists.")
+            flash("This username already exists.", "error_username")
             return redirect(url_for("auth.register"))
         if len(username) == 0:
-            flash("You must provide a username.")
+            flash("You must provide a username.", "error_username")
             return redirect(url_for("auth.register"))
+        if user_email:
+            flash("This email already exists.", "error_email")
+            return redirect(url_for("auth.register"))   
+        if confirm_email != email:
+            flash("The emails do not match!", "error_email_confirm")
+            return redirect(url_for("auth.register"))             
         if len(password) == 0:
-            flash("You must provide a password.")
+            flash("You must provide a password.", "error_pass")
             return redirect(url_for("auth.register"))
         if confirm_password != password:
-            flash("The passwords do not match!")
-            return redirect(url_for("auth.register"))
-        if len(name) == 0:
-            flash("You must provide your name.")
+            flash("The passwords do not match!", "error_pass_confirm")
             return redirect(url_for("auth.register"))
+        
 
         # checking password strength
         result = check_password_strength(password)
         if result:
-            flash("Password is not strong enough. Here are some suggestions: " + ", ".join(result))
+            flash("Password is not strong enough. Here are some suggestions: " + ", ".join(result), "error_pass")
             return redirect(url_for("auth.register"))
 
-        new_user = User(email=email, username=username, name=name)
+        new_user = User(
+            email=email, 
+            username=username, 
+            name=name,
+            grade=grade,
+            age=age
+        )
```

Comment:
I don't think it's necessary to include the age and grade upon user registration. Since the user has the option of leaving some fields blank (e.g. date of birth), we need to check if they are provided or not. If we do a "one size fits all" solution, this creates an invalid entry in our database because it cannot have DOB or other fields as "". 

An example would be the following. Assume the DOB is provided, the user addition to the database might look like `new_user = User(email=email, username=username, name=name, age=age)`. If the grade is provided but not the DOB, it might look like `new_user = User(email=email, username=username, name=name, grade=grade)`. The point is, we cannot have the incorrect type in our user database columns.

Reply:
I actually accounted for this by requiring any new users signing up to fill out all text fields, so this specific invalid entry situation will never occur as a result.

Basically, when a user visits the registration or login page, the form submit button will be grayed out and unclickable; only when all fields are filled out will it light up and be interactable again. Therefore, unless if a user somehow manages to sign up using some other means than the registration page, all new users will have the same number and types of columns filled out without any empty fields.